modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsConfigurationManager.java

```diff
@@ -0,0 +1,518 @@
+        GridInternalSubscriptionProcessor subscriptionProcessor,
+        GridCachePartitionExchangeManager exchange,
+        IgniteStatisticsRepository repo,
+        StatisticsGatherer gatherer,
+        IgniteThreadPoolExecutor mgmtPool,
+        Function<Class<?>, IgniteLogger> logSupplier
+    ) {
+        this.schemaMgr = schemaMgr;
+        log = logSupplier.apply(IgniteStatisticsConfigurationManager.class);
+        this.repo = repo;
+        this.mgmtPool = mgmtPool;
+        this.gatherer = gatherer;
+
+        subscriptionProcessor.registerDistributedMetastorageListener(new DistributedMetastorageLifecycleListener() {
+            @Override public void onReadyForRead(ReadableDistributedMetaStorage metastorage) {
+                distrMetaStorage = (DistributedMetaStorage)metastorage;
+
+                distrMetaStorage.listen(
+                    (metaKey) -> metaKey.startsWith(STAT_OBJ_PREFIX),
+                    (k, oldV, newV) -> {
+                        // Skip invoke on start node (see 'ReadableDistributedMetaStorage#listen' the second case)
+                        // The update statistics on start node is handled by 'scanAndCheckLocalStatistic' method
+                        // called on exchange done.
+                        if (!started)
+                            return;
+
+                        mgmtPool.submit(() -> {
+                            try {
+                                onChangeStatisticConfiguration(
+                                    (StatisticsObjectConfiguration)oldV,
+                                    (StatisticsObjectConfiguration)newV
+                                );
+                            }
+                            catch (Throwable e) {
+                                log.warning("Unexpected exception on change statistic configuration [old="
+                                    + oldV + ", new=" + newV + ']', e);
+                            }
+                        });
+                    }
+                );
+            }
+        });
+
+        exchange.registerExchangeAwareComponent(
+            new PartitionsExchangeAware() {
+                @Override public void onDoneAfterTopologyUnlock(GridDhtPartitionsExchangeFuture fut) {
+                    started = true;
+
+                    scanAndCheckLocalStatistic(fut.topologyVersion());
+                }
+            }
+        );
+
+        schemaMgr.registerDropColumnsListener(this::onDropColumns);
+        schemaMgr.registerDropTable(this::onDropTable);
+    }
+
+    /** */
+    private void scanAndCheckLocalStatistic(AffinityTopologyVersion topVer) {
+        mgmtPool.submit(() -> {
+            try {
+                distrMetaStorage.iterate(STAT_OBJ_PREFIX, (k, v) ->
+                    checkLocalStatistics((StatisticsObjectConfiguration)v, topVer));
+            }
+            catch (IgniteCheckedException e) {
+                log.warning("Unexpected exception on check local statistic on start", e);
```

Comment:
Not only on start but after each topology change. It could be too heavy to load all the stats each time. What if we'll track partitions from wich local statistics were aggregated? To just compare it with a new list of local primary partitions.